Support for version constraint wildcards

beornf opened the pull request:
I discovered you started a library after commenting on my PR at https://github.com/hashicorp/go-version/pull/49. I've written this PR which adds support for version wildcards and passes all https://github.com/Masterminds/semver test cases.

This also addresses a few issues:
- Repeated calls to Constraints `Check` will mutate the object in `PessimisticBump`, `TildeBump` and `CaretBump`
- Adding a prerelease check for all constraint functions
- Empty string is treated the same as wildcards

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/aquasecurity/go-version?pullRequest=3) <br/>All committers have signed the CLA.

beornf committed:
Support for version constraint wildcards

beornf commented:
I have reworked this PR so there are no breaking changes to existing test cases. I have also added support for range constraints for example, `1 - 2`.

knqyf263 commented on pkg/version/constraint.go:
```diff
@@ -90,15 +102,43 @@ func NewConstraints(v string) (Constraints, error) {
-	v, err := Parse(m[2])
-	if err != nil {
-		return Constraint{}, xerrors.Errorf("version parse error (%s): %w", m[2], err)
+	var segments []part.Part
+	for _, str := range strings.Split(m[3], ".") {
+		segments = append(segments, part.NewPart(str))
+	}
+
+	v := Version{
+		segments:   segments,
+		preRelease: part.NewParts(m[6]),
+		original:   c,
```
I'm curious why `Parse` doesn't fit here.

beornf replied:
If we use `Parse` from `version.go` here that means a valid version would be `1.2.x`. So instead it's necessary to parse the segments here to create a valid "constraint" version.

knqyf263 reviewed (commented):
We're sorry to miss this PR.
